File: lib/cnf.ml
(** cnf.ml — CNF conversion pipeline for the Resolution Engine.
    
    Transforms a formula into a set of clauses (sets of literals).
    Pipeline: Negate -> Simplify -> NNF -> Skolemize -> Drop Forall -> Distribute -> Clausify. *)

open Types
open Formula

(** Convert a formula to Negation Normal Form (NNF).
    Push negations inwards until they are only in front of predicates. *)
let rec nnf = function
  | Not (Not a) -> nnf a
  | Not (And (a, b)) -> Or (nnf (Not a), nnf (Not b))
  | Not (Or (a, b)) -> And (nnf (Not a), nnf (Not b))
  | Not (Implies (a, b)) -> And (nnf a, nnf (Not b))
  | Not (Iff (a, b)) -> Or (And (nnf a, nnf (Not b)), And (nnf (Not a), nnf b))
  | Not (Forall (x, a)) -> Exists (x, nnf (Not a))
  | Not (Exists (x, a)) -> Forall (x, nnf (Not a))
  | And (a, b) -> And (nnf a, nnf b)
  | Or (a, b) -> Or (nnf a, nnf b)
  | Implies (a, b) -> Or (nnf (Not a), nnf b)
  | Iff (a, b) -> And (Or (nnf (Not a), nnf b), Or (nnf (Not b), nnf a))
  | Forall (x, a) -> Forall (x, nnf a)
  | Exists (x, a) -> Exists (x, nnf a)
  | f -> f

(** Skolemize a formula in NNF.
    Replace existentially quantified variables with Skolem functions of
    the universally quantified variables in scope. *)
let skolemize f =
  let rec skolem (stats : int ref) (univ_vars : term list) = function
    | Top | Bot | Pred _ as f -> f
    | Not a -> Not (skolem stats univ_vars a)
    | And (a, b) -> And (skolem stats univ_vars a, skolem stats univ_vars b)
    | Or (a, b) -> Or (skolem stats univ_vars a, skolem stats univ_vars b)
    | Forall (x, a) -> Forall (x, skolem stats (Var x :: univ_vars) a)
    | Exists (x, a) ->
      let sk_name = "sk_" ^ string_of_int !stats in
      stats := !stats + 1;
      let sk_term = Func (sk_name, List.rev univ_vars) in
      skolem stats univ_vars (subst [(x, sk_term)] a)
    | Implies _ | Iff _ -> failwith "skolemize: expected NNF (no implies/iff)"
  in
  skolem (ref 0) [] f

(** Drop universal quantifiers (implicit quantification). *)
let rec drop_forall = function
  | Forall (_, a) -> drop_forall a
  | And (a, b) -> And (drop_forall a, drop_forall b)
  | Or (a, b) -> Or (drop_forall a, drop_forall b)
  | f -> f

(** Distribute Or over And to reach Conjunctive Normal Form.
    (A /\ B) \/ C  =>  (A \/ C) /\ (B \/ C) *)
let rec distribute = function
  | And (a, b) -> And (distribute a, distribute b)
  | Or (And (a, b), c) ->
    let a', b', c' = distribute a, distribute b, distribute c in
    And (distribute (Or (a', c')), distribute (Or (b', c')))
  | Or (c, And (a, b)) ->
    let a', b', c' = distribute a, distribute b, distribute c in
    And (distribute (Or (c', a')), distribute (Or (c', b')))
  | Or (a, b) ->
    let a', b' = distribute a, distribute b in
    (* We must re-check if distribution is possible after recursing *)
    (match a', b' with
     | And _, _ | _, And _ -> distribute (Or (a', b'))
     | _ -> Or (a', b'))
  | f -> f

(** Convert a CNF formula to a list of clauses (disjunctions of literals). *)
let rec clausify = function
  | And (a, b) -> clausify a @ clausify b
  | f ->
    let rec literals = function
      | Or (a, b) -> literals a @ literals b
      | Pred (p, args) -> [Pos (p, args)]
      | Not (Pred (p, args)) -> [Neg (p, args)]
      | Top -> [] (* Should have been simplified *)
      | Bot -> []
      | _ -> failwith "clausify: expected literals (NNF)"
    in
    [literals f]

(** The full transformation pipeline. *)
let to_clauses f =
  f
  |> simplify
  |> nnf
  |> skolemize
  |> drop_forall
  |> distribute
  |> clausify

File: python/cnf.py
"""
cnf.py — CNF conversion pipeline for the Resolution Engine.
Ported from cnf.ml and formula.ml.
"""

from fol_types import (
    Term, Variable, Constant, Function,
    Formula, Atom, Negation, BinaryOp, Connective,
    QuantifiedFormula, Quantifier
)
from formula_utils import simplify, nnf, variant, free_vars

def skolemize(f: Formula) -> Formula:
    """Skolemize a formula in NNF."""
    counter = [0]
    
    def sk(univ_vars: list[Term], f: Formula) -> Formula:
        if isinstance(f, Atom):
            return f
        if isinstance(f, Negation):
            return Negation(sk(univ_vars, f.formula))
        if isinstance(f, BinaryOp):
            return BinaryOp(f.connective, sk(univ_vars, f.left), sk(univ_vars, f.right))
        if isinstance(f, QuantifiedFormula):
            if f.quantifier == Quantifier.FORALL:
                return QuantifiedFormula(Quantifier.FORALL, f.variable, sk(univ_vars + [f.variable], f.body))
            if f.quantifier == Quantifier.EXISTS:
                sk_name = f"sk_{counter[0]}"
                counter[0] += 1
                sk_term = Function(sk_name, tuple(univ_vars))
                # Substitute bound variable with skolem term
                body_prime = f.instantiate(sk_term)
                return sk(univ_vars, body_prime)
        return f

    return sk([], f)

def drop_forall(f: Formula) -> Formula:
    """Drop universal quantifiers (implicit quantification)."""
    if isinstance(f, QuantifiedFormula):
        if f.quantifier == Quantifier.FORALL:
            return drop_forall(f.body)
    if isinstance(f, BinaryOp):
        return BinaryOp(f.connective, drop_forall(f.left), drop_forall(f.right))
    return f

def distribute(f: Formula) -> Formula:
    """Distribute Or over And to reach Conjunctive Normal Form."""
    if isinstance(f, BinaryOp):
        if f.connective == Connective.AND:
            return BinaryOp(Connective.AND, distribute(f.left), distribute(f.right))
        if f.connective == Connective.OR:
            l = distribute(f.left)
            r = distribute(f.right)
            
            if isinstance(l, BinaryOp) and l.connective == Connective.AND:
                return BinaryOp(Connective.AND, distribute(BinaryOp(Connective.OR, l.left, r)), 
                                               distribute(BinaryOp(Connective.OR, l.right, r)))
            if isinstance(r, BinaryOp) and r.connective == Connective.AND:
                return BinaryOp(Connective.AND, distribute(BinaryOp(Connective.OR, l, r.left)), 
                                               distribute(BinaryOp(Connective.OR, l, r.right)))
            return BinaryOp(Connective.OR, l, r)
    return f

# For resolution, we need literals and clauses
# A Literal is a Pos or Neg predicate
# Since we already have Atom and Negation(Atom), we can use those or define a specific class.
# The user asked to define Resolution-specific types in resolution.py.
# So I'll just return lists of lists of formulas for now, and resolution.py will wrap them.

def clausify(f: Formula) -> list[list[Formula]]:
    """Convert a CNF formula to a list of clauses (disjunctions of literals)."""
    if isinstance(f, BinaryOp) and f.connective == Connective.AND:
        return clausify(f.left) + clausify(f.right)
    
    def literals(fm: Formula) -> list[Formula]:
        if isinstance(fm, BinaryOp) and fm.connective == Connective.OR:
            return literals(fm.left) + literals(fm.right)
        if isinstance(fm, Atom):
            if fm.predicate in ("⊤", "⊥"): return []
            return [fm]
        if isinstance(fm, Negation) and isinstance(fm.formula, Atom):
            return [fm]
        return []

    return [literals(f)]

def to_clauses(f: Formula) -> list[list[Formula]]:
    """The full transformation pipeline."""
    return clausify(distribute(drop_forall(skolemize(nnf(simplify(f))))))
